controllers/C_Document.class.php

```diff
@@ -533,48 +534,85 @@ function retrieve_action($patient_id = "", $document_id, $as_file = true, $origi
+                header("Content-Type: " . $d->get_mimetype());
+                header("Content-Length: " . strlen($filetext));
+                echo $filetext;
             }
             exit;//exits only if file download from CouchDB is successfull.
         }
+        if ($couch_docid && $couch_revid) {
+            //special case when retrieving a document from couchdb that has been converted to a jpg and not directly referenced in openemr documents table
+            //try to convert it if it has not yet been converted
+            $couch = new CouchDB();
+            $data = array($GLOBALS['couchdb_dbase'], "converted_".$couch_docid);
+            $resp = $couch->retrieve_doc($data);
+            $content = $resp->data;
+            if ($content == '') {
+                $couchM = new CouchDB();
+                $dataM = array($GLOBALS['couchdb_dbase'], $couch_docid);
+                $respM = $couchM->retrieve_doc($dataM);
+                $contentM = $respM->data;
+                if ($contentM=='' && $GLOBALS['couchdb_log']==1) {
+                    $log_content = date('Y-m-d H:i:s')." ==> Retrieving document\r\n";
+                    $log_content = date('Y-m-d H:i:s')." ==> URL: ".$url."\r\n";
+                    $log_content .= date('Y-m-d H:i:s')." ==> CouchDB Document Id: ".$couch_docid."\r\n";
+                    $log_content .= date('Y-m-d H:i:s')." ==> CouchDB Revision Id: ".$couch_revid."\r\n";
+                    $log_content .= date('Y-m-d H:i:s')." ==> Failed to fetch document content from CouchDB.\r\n";
+                    $log_content .= date('Y-m-d H:i:s')." ==> Will try to download file from HardDisk if exists.\r\n\r\n";
+                    $this->document_upload_download_log($d->get_foreign_id(), $log_content);
+                    die(xlt("File retrieval from CouchDB failed"));
+                }
+                $from_file_tmp_name = tempnam($GLOBALS['temporary_files_dir'], "oer");
+                file_put_contents($from_file_tmp_name, base64_decode($contentM));
+                $to_file_tmp_name = tempnam($GLOBALS['temporary_files_dir'], "oer") . ".jpg";
+                exec("convert -density 200 " . escapeshellarg($from_file_tmp_name) . " -append -resize 850 " . escapeshellarg($to_file_tmp_name));
+                unlink($from_file_tmp_name);
+                if (is_file($to_file_tmp_name)) {
+                    $couchI = new CouchDB();
+                    $json = json_encode(base64_encode(file_get_contents($to_file_tmp_name)));
+                    $couchdata = array($GLOBALS['couchdb_dbase'], "converted_" . $couch_docid, $d->get_foreign_id(), "", "image/jpeg", $json);
+                    $couchI->check_saveDOC($couchdata);
+                    unlink($to_file_tmp_name);
+                } else {
+                    echo "<b>NOTE</b>: " . xlt('Document') . "'" . text($fname) . "' " . xlt('cannot be converted to JPEG. Perhaps ImageMagick is not installed?') . "<br><br>";
```

Comment:
TODO
Likely need to deal with this message differently now that brought in documents class